src/common/utils/oidc/secret.go

```diff
@@ -60,80 +62,91 @@ func (dm *defaultManager) getEncryptKey() (string, error) {
-		return fmt.Errorf("user is not onboarded as OIDC user")
+		return nil, fmt.Errorf("user is not onboarded as OIDC user")
 	}
 	key, err := dm.getEncryptKey()
 	if err != nil {
-		return fmt.Errorf("failed to load the key for encryption/decryption： %v", err)
+		return nil, fmt.Errorf("failed to load the key for encryption/decryption： %v", err)
 	}
 	plainSecret, err := utils.ReversibleDecrypt(oidcUser.Secret, key)
 	if err != nil {
-		return fmt.Errorf("failed to decrypt secret from DB: %v", err)
+		return nil, fmt.Errorf("failed to decrypt secret from DB: %v", err)
 	}
 	if secret != plainSecret {
-		return verifyError(errors.New("secret mismatch"))
-	}
-	return dm.VerifyToken(ctx, oidcUser)
-}
-
-// VerifyToken checks the expiration of the token in the model, (we'll only do expiration checks b/c according to spec,
-// the response may not have ID token:
-// https://openid.net/specs/openid-connect-core-1_0.html#RefreshTokenResponse
-// and it will try to refresh the token
-// if it's expired, if the refresh is successful it will persist the token and consider the verification successful.
-func (dm *defaultManager) VerifyToken(ctx context.Context, user *models.OIDCUser) error {
-	if user == nil {
-		return verifyError(fmt.Errorf("input user is nil"))
-	}
-	key, err := dm.getEncryptKey()
-	if err != nil {
-		return fmt.Errorf("failed to load the key for encryption/decryption： %v", err)
+		return nil, verifyError(errors.New("secret mismatch"))
 	}
-	tokenStr, err := utils.ReversibleDecrypt(user.Token, key)
+	tokenStr, err := utils.ReversibleDecrypt(oidcUser.Token, key)
 	if err != nil {
-		return verifyError(err)
+		return nil, verifyError(err)
 	}
 	token := &Token{}
 	err = json.Unmarshal(([]byte)(tokenStr), token)
 	if err != nil {
-		return verifyError(err)
+		return nil, verifyError(err)
 	}
-	log.Debugf("Token string for verify: %s", tokenStr)
-	if token.Expiry.After(time.Now()) {
-		return nil
+	// Best effort to get the most up-to-date ID token
+	newToken, err := RefreshToken(ctx, token)
+	if err != nil && token.Expiry.After(time.Now()) {
+		return nil, fmt.Errorf("failed to refresh token when it's expired, error: %v", err)
 	}
-	log.Info("Token string has expired, refreshing...")
-	t, err := RefreshToken(ctx, token)
 	if err != nil {
-		return verifyError(err)
+		log.Warningf("failed to refresh token, error: %v reusing the old one.", err)
+		newToken = token
 	}
-	tb, err := json.Marshal(t)
+	if newToken.IDToken == "" {
+		log.Warningf("reusing the ID token of the old ID token.")
+		newToken.IDToken = token.IDToken
+	}
+
+	// Persist the token to DB.  Assuming there's low possibility for concurrent refresh scenario
+	tb, err := json.Marshal(newToken)
 	if err != nil {
 		log.Warningf("Failed to encode the refreshed token, error: %v", err)
 	}
 	encToken, _ := utils.ReversibleEncrypt(string(tb), key)
-	user.Token = encToken
-	err = dao.UpdateOIDCUser(user)
+	oidcUser.Token = encToken
+	err = dao.UpdateOIDCUser(oidcUser)
```

Comment:
The update of oidc user  here is not needed when token is not refreshed.

```go
// RefreshToken refreshes the token passed in parameter, and return the new token.
func RefreshToken(ctx context.Context, token *Token) (*Token, error) {
	oauth, err := getOauthConf()
	if err != nil {
		log.Errorf("Failed to get OAuth configuration, error: %v", err)
		return nil, err
	}
	setting := provider.setting.Load().(models.OIDCSetting)
	ctx = clientCtx(ctx, setting.VerifyCert)
	ts := oauth.TokenSource(ctx, &token.Token)
	t, err := ts.Token()
	if err != nil {
		return nil, err
	}

	it, ok := t.Extra("id_token").(string)
	if !ok {
		log.Debugf("id_token not exist in refresh response")
	}
	return &Token{Token: *t, IDToken: it}, nil
}
```

`t, err := ts.Token()` may return old token when it's still valid.


```go
// Token returns the current token if it's still valid, else will
// refresh the current token (using r.Context for HTTP client
// information) and return the new one.
func (s *reuseTokenSource) Token() (*Token, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.t.Valid() {
		return s.t, nil
	}
	t, err := s.new.Token()
	if err != nil {
		return nil, err
	}
	s.t = t
	return t, nil
}
```

Reply:
No longer valid after the rework.